### PR TITLE
feat: Enable deferred unregistering of shared memory regions after inference

### PR DESCRIPTION
**Pull request by pskiran1**
#### What does the PR do?
Currently, an error is returned if an attempt is made to unregister a shared memory region that is in use during inference (introduced in #7567). This PR introduces the following enhancement:

- When the unregister API is called for a region still in use, the API will now mark the region for pending unregistration. The server will automatically remove the region once all in-progress inferences utilizing it have completed.
- In the interim, users can query the server to check if the region is still retained, allowing for safe cleanup once the server has fully released the region.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [ ] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [x] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Test plan:

- CI Pipeline ID: 19695563

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
- closes GitHub issue: #7688

### CODE REVIEW

File: qa/L0_cuda_shared_memory/cuda_shared_memory_test.py [hunk 1]
```diff
@@ -401,74 +401,31 @@ def test_infer_byte_size_out_of_bound(self):
...
         second_client = httpclient.InferenceServerClient("localhost:8000", verbose=True)
 
-        with self.assertRaises(InferenceServerException) as ex:
-            second_client.get_cuda_shared_memory_status("input0_data")
-        self.assertIn(
-            "Unable to find cuda shared memory region: 'input0_data'",
-            str(ex.exception),
-        )
+        status = second_client.get_cuda_shared_memory_status()
+        self.assertEqual(len(status), 0)
```
[the remark this answers is not in the document]
Comment (pskiran1) [reply qa/L0_cuda_shared_memory/cuda_shared_memory_test.py]: Thank you. I modified `_test_shm_not_found` to check the shared memory status by name, similar to the previous implementation. I also cleaned up the CUDA shared memory test similar to your recent change #7729.

File: src/http_server.h [hunk 2]
```diff
@@ -272,13 +272,34 @@ class HTTPAPIServer : public HTTPServer {
...
+        auto shm_name = shm_regions_info_.back()->name_;
+        auto shm_memory_type = shm_regions_info_.back()->kind_;
+        auto awaiting_unregister =
+            shm_regions_info_.back()->awaiting_unregister_;
+
+        // Delete shared_ptr to decrement reference count
+        shm_regions_info_.pop_back();
+
+        if (awaiting_unregister) {
+          auto err = shm_manager_->Unregister(shm_name, shm_memory_type);
```
[the remark this answers is not in the document]
Comment (pskiran1) [reply src/http_server.h]: Done. Added a `nullptr` check on `shm_manager_` before calling `Unregister()` and log verbose if it is `nullptr`.

File: src/shared_memory_manager.h [hunk 3]
```diff
@@ -72,6 +72,7 @@ class SharedMemoryManager {
     void* mapped_addr_;
     TRITONSERVER_MemoryType kind_;
     int64_t device_id_;
+    bool awaiting_unregister_;
```
Comment (nnshah1) [inline src/shared_memory_manager.h:75]: question: could this be done with smart pointer referencing counting instead of an explicit flag?

Comment (pskiran1) [reply src/shared_memory_manager.h:75]: @nnshah1, yes, I believe we can implement this with `weak_ptr`.
@GuanLuo has previously provided input on an alternative approach using `weak_ptr` and also discussed the current method.
Here’s a summary of the `weak_ptr` approach:
Currently, the shared memory (shm) manager uses a map of shared_ptr for shm_info. When the unregister API is called on a region still in use, we can convert it to a `weak_ptr` to decrease the reference count. We would track these in a separate `weak_ptr map` until fully unregistered, preventing new regions with the same name and allowing status check.

When the last reference to the `shm_info` `shared_ptr` is released (after inference completes), the `shared_ptr` destructor will trigger. At that point, we can invoke the unregister logic and remove the associated `weak_ptr` from the second map.

To support this approach, we would likely need to modify most of shm manager methods, such as register, status, unregister, and unregisterHelper, etc. to work with the `weak_ptr map`.

For simplicity, I proceeded with the current approach. Please let me know if we’d like to explore the `weak_ptr` approach, and I can work on it. Happy to connect and discuss if any optimizations are needed.
Thank you.

Comment (nnshah1) [reply src/shared_memory_manager.h:75]: Thanks for the detailed response - it was just that it seemed like a lot of similarity with reference counting and shared pointers - so was hoping it was something quick. But if this is the more practical short term - no issue, 

@GuanLuo - would we want to pre-emptively put a refactor story in? (only if you think its a near term thing that needs to be addressed) -

File: qa/L0_cuda_shared_memory/cuda_shared_memory_test.py [hunk 4]
```diff
@@ -397,171 +409,222 @@ def test_infer_byte_size_out_of_bound(self):
...
+
+        # Unregister all should not result in an error.
+        # If shared memory regions are in use, they will be marked and unregistered after the inference is completed.
+        with httpclient.InferenceServerClient(
+            "localhost:8000", verbose=True
+        ) as second_client:
+            second_client.unregister_cuda_shared_memory()
+
+        # Number of shared memory regions should be the same as the inference is not completed yet
+        self._test_shm_found(shm_names)
```
Comment (GuanLuo) [inline qa/L0_cuda_shared_memory/cuda_shared_memory_test.py:427]: So the region will still be visible between unregister and completion of the inference, what would happen if another inference request is sent during that time, trying to reuse the unregistered region? Can you add a test for this case to demonstrate the behavior?

File: src/shared_memory_manager.h (hunk 3 above)
Comment (GuanLuo) [reply src/shared_memory_manager.h:75]: I have raised the point that using smart pointer is a cleaner way to address this problem early on for this development, so you can have a refactor story as it is just a right thing to do. However, I believe that it is just going to be another form of TODOs at this point and what is in this PR will remain unchanged for a long time.

File: qa/L0_cuda_shared_memory/cuda_shared_memory_test.py (hunk 4 above)
Comment (pskiran1) [reply qa/L0_cuda_shared_memory/cuda_shared_memory_test.py]: Yes, the shm will remain accessible between unregister and the completion of the inference. When we run another inference request that reuses the unregistered region, the shm region will be extended and unregistered after the second request, as we check the [reference count](https://github.com/triton-inference-server/server/pull/7743/files#diff-b2df344fcbed5fc3756d51cd3ea194ac9edecbcc7a6f40022642520c92cf8f78L688) before unregister. Added a test case to verify this.

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**